sktime/forecasting/compose/_autoregressive.py

```diff
@@ -0,0 +1,303 @@
+        self.forecaster = forecaster
+        self.horizon_length = horizon_length
+        self.aggregate_method = (
+            aggregate_method if aggregate_method is not None else np.mean
+        )
+
+        super().__init__()
+
+    def _fit(self, y, X=None, fh=None):
+        # ignore the fh provided fit
+        _fh = ForecastingHorizon(range(1, self.horizon_length + 1))
+
+        self.forecaster.fit(y=y, X=X, fh=_fh)
+
+    def _predict(self, fh, X=None, y=None):
+        # use fh to find the maximum length to forecast regressively
+        max_fh = max(
+            *fh.to_relative(self._cutoff)._values,
+            self.horizon_length,
+        )
+
+        hist_y = self._y if y is None else y
+        hist_y = deepcopy(hist_y)  # make copy because we are going to add values to it
+
+        # calculate initial length of hist_y
+        if not isinstance(hist_y.index, pd.MultiIndex):
+            initial_length = len(hist_y.index)
+        else:
+            initial_length = len(hist_y.index.levels[-1])
+
+        if y is None and X is not None:
+            # special case of non global forecasting
+            # here X contains only (future) values
+            # however for the interface, X should contain (hist + future) values
+            # since this is non-global forecasting, we can use hist from self._X
+            X = pd.concat([self._X, X])
+
+        for i in range(max_fh - self.horizon_length + 1):
+            # truncate to the needed lengths
+            _y = self._truncate(hist_y, end_index=initial_length + i)
+            _x = self._truncate(X, end_index=initial_length + i + self.horizon_length)
+
+            # forecast for self.horizon_length
+            preds = self.forecaster.predict(y=_y, X=_x)
+
+            # use predicted values as history for next iteration
+            hist_y = self._concate_preds(hist_y, preds)
+
+        # collect forecasting points from hist_y
+        # code works for all indexes
+        absolute_horizons = fh.to_absolute_index(self.cutoff)
+        dateindex = hist_y.index.get_level_values(-1).map(
+            lambda x: x in absolute_horizons
+        )
+        preds = hist_y.loc[dateindex]
+
+        return preds
+
+    def _truncate(self, data, end_index):
+        # keep rows till this index in data
+
+        if data is None:
+            # when there is no X (exogenous data)
+            return None
+
+        if not isinstance(data.index, pd.MultiIndex):
+            return data.head(end_index)
+
+        # truncating for Multi-Index
+        # levels other than the timestamps (innermost/base level)
+        groupby_levels = list(range(len(data.index.names) - 1))
+        data = data.groupby(
+            level=groupby_levels,
+            group_keys=True,  # to keep stable conversion with pandas<2.0.0
+        ).apply(lambda x: x.head(end_index))
+        data.index = data.index.droplevel(groupby_levels)
```

Comment:
Can you please give an example of why/where this line is required?

Reply:
```py
>>> import pandas as pd
>>> import numpy as np
>>> 
>>> # preparing data
>>> 
>>> date_range = pd.date_range(start='2000-01-01', end='2000-01-5')
>>> h0_values = [f'h0_{i}' for i in range(90, 92)]
>>> index = pd.MultiIndex.from_product([h0_values, date_range], names=['h0', 'time'])
>>> data = np.random.randn(len(index), 2) + 5
>>> data = pd.DataFrame(data, index=index, columns=['c2', 'c3'])
>>> data
                        c2        c3
h0    time                          
h0_90 2000-01-01  4.621762  5.023309
      2000-01-02  7.370289  4.904299
      2000-01-03  5.042544  5.857899
      2000-01-04  5.835349  5.689656
      2000-01-05  4.159186  4.608287
h0_91 2000-01-01  6.273114  5.328466
      2000-01-02  5.159423  3.774948
      2000-01-03  3.651350  5.264665
      2000-01-04  4.694018  5.348987
      2000-01-05  5.685372  5.328585
>>> 
>>> # code from estimator
>>> 
>>> groupby_levels = list(range(len(data.index.names) - 1))
>>> groupby_levels
[0]
>>> 
>>> data = data.groupby(
...     level=groupby_levels,
...     group_keys=True,  # to keep stable conversion with pandas<2.0.0
... ).apply(lambda x: x.head(2))
>>> data
                              c2        c3
h0    h0    time                          
h0_90 h0_90 2000-01-01  4.621762  5.023309
            2000-01-02  7.370289  4.904299
h0_91 h0_91 2000-01-01  6.273114  5.328466
            2000-01-02  5.159423  3.774948
>>> 
>>> data.index = data.index.droplevel(groupby_levels)
>>> data
                        c2        c3
h0    time                          
h0_90 2000-01-01  4.621762  5.023309
      2000-01-02  7.370289  4.904299
h0_91 2000-01-01  6.273114  5.328466
      2000-01-02  5.159423  3.774948
```

Reply:
```pycon
>>> 
>>> pd.__version__
'1.5.3'
>>> 
>>> a = data.groupby(level=groupby_levels, group_keys=True).apply(lambda x: x.head(2))
>>> b = data.groupby(level=groupby_levels).apply(lambda x: x.head(2))
>>> 
>>> pd.testing.assert_frame_equal(a, b)
>>> 
>>> a
                              c2        c3
h0    h0    time                          
h0_90 h0_90 2000-01-01  4.555509  4.708377
            2000-01-02  5.723046  6.357103
h0_91 h0_91 2000-01-01  7.250341  5.366647
            2000-01-02  3.496937  6.019925
>>> 
```

Why do we need `group_keys` to add the index, only to drop it later?

Reply:
Also question to @fkiraly and @benHeid : we are keeping pandas v1 compatibility for existing estimators to support users who can't quickly upgrade. But do we need to ensure v1 compatibility for every new estimator getting added as well? Will that not only increase the amount of "old" style stuff?

Reply:
`group_keys` is not adding the index, index is added anyways for any pandas version, `group_keys` was added here just to fix the failing test-unix-pandas1 tests on this [workflow](https://github.com/sktime/sktime/actions/runs/10235760957)

Reply:
I tested on my local environment yesterday, and found that indices are duplicated only when I add group keys with v2. I dropped it, it was fine and I tested same with v1 (above snippet) and it seems same as your intended result.